### ISSUE TITLE
ANSI Escape Sequences Terminate/Bork Logs

### ISSUE THREAD

**Issue by schmidtw** [concourse/concourse#7464]
## ANSI Escape Sequences Terminate/Bork Logs

We're using concourse as a driver for building vm images & in the stream of output from those, several [ANSI escape sequences](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797) are present.  Largely I don't care if the output looks bad, but some of the control codes are permanently breaking the output (at least on the web-ui).

```
2021/08/20 08:05:50 packer-builder-qemu plugin: Qemu stdout: [^[[0;32m  OK  ^[[0m] Started Hardware RNG Entropy Gatherer Wake threshold service.
2021/08/20 08:05:50 packer-builder-qemu plugin: Qemu stdout: [^[[0;32m  OK  ^[[0m] Started Anaconda NetworkManager configuration.
2021/08/20 08:05:50 packer-builder-qemu plugin: Qemu stdout: ^[[r^[[m^[(B^[[?1003l^[[?1006l^[[?2004l^[[1;1H^[[1;24r^[[1;1H
2021/08/20 08:05:50 packer-builder-qemu plugin: Qemu stdout:
```

The collection of escape sequences above are the last thing ever seen in the concourse web-ui log.  I would expect concourse to filter all the escape sequences that it can't handle.

## Steps to reproduce

Output the following (replaceing ^[ with the control code `0x1b`) and the web-ui's logs appear to hang until the build completes.
```
2021/08/20 08:05:50 packer-builder-qemu plugin: Qemu stdout: ^[[r^[[m^[(B^[[?1003l^[[?1006l^[[?2004l^[[1;1H^[[1;24r^[[1;1H
```

## Expected results

The webui just ignores all the control sequences except the ones it can handle (today this looks like just the colors).

## Actual results

The webui just stops working.

## Triaging info

* Concourse version: ` v6.7.6`
* Browser (if applicable): Firefox
* Did this used to work? Unsure ... probably not.

**Comment by taylorsilva** [comment]
Reproducible pipeline:
```yaml
---
jobs:
- name: job
  public: true
  plan:
  - task: simple-task
    config:
      platform: linux
      image_resource:
        type: mock
        source: { mirror_self: true }
      run:
        path: bash
        args:
        - -c
        - |
          echo "hey there"
          echo -e 'out: \033[r\033[m\033(B\033[?1003l\033[?1006l\033[?2004l\033[1;1H\033[1;24r\033[1;1H '
          echo "bye now"
```

<img width="330" alt="image" src="https://user-images.githubusercontent.com/4583837/132406472-751c6dc4-0aab-464b-ab22-8de78729741c.png">

**Comment by taylorsilva** [comment]
This escape sequence seems pretty nasty. Even my terminal (iterm2) renders it badly:

https://user-images.githubusercontent.com/4583837/132408983-0de34f58-6fa8-4f96-9d47-6f60469abb32.mov

I'm not sure this is actually a "bug" we can fix, given it renders badly in my terminal as well. Do you have an example that renders correctly in a terminal but not in the web ui?

and I just realized you said the terminal hangs for you. For me the terminal is not hanging, output still appears, though everything is generally wonky now:

```yaml
---
jobs:
- name: job
  public: true
  plan:
  - task: simple-task
    config:
      platform: linux
      image_resource:
        type: mock
        source: { mirror_self: true }
      run:
        path: bash
        args:
        - -c
        - |
          echo "hey there"
          echo -e 'out: \033[r\033[m\033(B\033[?1003l\033[?1006l\033[?2004l\033[1;1H\033[1;24r\033[1;1H '
          echo "sleep for 3s"
          sleep 3
          echo "done sleeping"
```

Like I said though, doing this through fly also messes up my local terminal until I do a `clear` to reset everything. Until I do that my terminal jumps around for anything with output that's larger than the terminal size.

https://user-images.githubusercontent.com/4583837/132409924-54949e3f-5e5b-4a92-8516-17c7df01a9f3.mov

---

To recap, I think this initial example is bad for any client trying to render it, even real terminals. It also doesn't not hang (in my case), later output still renders. Do you have an example that renders in a terminal (like iterm) correctly but not in the web ui?

**Comment by schmidtw** [comment]
It works fine with `gnome-terminal` as well as whatever terminal is used during Centos8/Rocky8 installation.  That's where I got this bundle of joy.  After you've entered all the details (or preconfigured the kickstart file) this sequence happens and it clears the screen and continues.

The issue appears to be that the component(s) that are processing the log strings are smart enough to find the ANSI color instructions, but not smart enough to ignore everything it can't or shouldn't handle.  The output from the program is clearly written for a terminal window, (it resets the page & moves the cursor around a bunch) but that's our only option for logs ... we're installing a stock linux distribution. :)

I'll see if I can make a video of installing Rocky8 in a window if you think that's valuable.

**Comment by taylorsilva** [comment]
Looking through https://github.com/vito/elm-ansi/blob/master/src/Ansi.elm it looks like the codes that clear the screen and move the cursor are supported https://github.com/vito/elm-ansi/blob/05f89150e8bce99ae863cccda83e636564a64561/src/Ansi.elm#L45-L52

I guess if we have a specific list of codes we should ignore we could make that change. I don't thinking ignoring the very general escape codes would be a good idea though as it might lead to **really** wonky looking output. Maybe https://github.com/vito/elm-ansi/blob/05f89150e8bce99ae863cccda83e636564a64561/src/Ansi.elm#L51 could be ignored though?

**Comment by schmidtw** [comment]
I'm thinking about how to make a change for this workable for most folks.  Most things, most of the time don't have any of these problems & probably like the cursor manipulation tricks (because they make download status bars for example).

For folks that run into stuff like what I've run into it would be great if there was a way to indicate "ignore/consume/strip" any of the cursor moving features from lines from lines [43](https://github.com/vito/elm-ansi/blob/05f89150e8bce99ae863cccda83e636564a64561/src/Ansi.elm#L43) through [54](https://github.com/vito/elm-ansi/blob/05f89150e8bce99ae863cccda83e636564a64561/src/Ansi.elm#L54).

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.

**Comment by analytically** [comment]
@taylorsilva I'd like to focus a little more on this, can you add to planning again? Unhandled codes should at least not bork output

**Comment by analytically** [comment]
This should now be fixed by https://github.com/vito/elm-ansi/pull/16
